src/main/java/jenkins/branch/BranchProjectFactory.java

```diff
@@ -181,6 +181,35 @@ public void setRevisionHash(P project, SCMRevision revision) throws IOException
         file.write(revision);
     }
 
+    /**
+     * Gets the {@link SCMRevision} that the project was last seen for.
+     *
+     * @param project the project.
+     * @return the {@link SCMRevision} of the last seen.
+     */
+    @CheckForNull
+    public SCMRevision getLasSeenRevision(P project) {
+        XmlFile file = new XmlFile(new File(project.getRootDir(), "scm-last-seen-revision-hash.xml"));
```

Comment:
How often is this being called? Once per scan per folder? once per scan per child per folder?

File operations like this makes me a bit nervous since the IO performance of Jenkins is already a bit crappy and multiplication can produce powerful numbers :)

Could it be stored in the project xml as normal or why was that approach not chosen?

Reply:
The reason we store the last built revision in the file was so that the config history plugin doesn't see config changes that are not. This should only be called once per branch per event to the branch or per scan of the repo